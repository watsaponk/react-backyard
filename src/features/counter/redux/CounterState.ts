export default interface CounterState {
	value: number
	magicNumber: number
}
