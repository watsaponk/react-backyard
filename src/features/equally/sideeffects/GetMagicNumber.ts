export default function GetMagicNumber(): Promise<number> {
	return new Promise(resolve =>
		setTimeout(() => {
			const base = Math.floor(Math.random() * 6)
			const pow = Math.random() < 0.5 ? -1 : 1
			resolve(base * pow)
		}, 1000)
	)
}
