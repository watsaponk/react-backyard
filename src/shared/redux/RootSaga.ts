import { fork, ForkEffect } from 'redux-saga/effects'
import equallySaga from '../../features/equally/EquallySaga'
import userListSaga from '../../features/userlist/UserListSaga'

export default function* watchRootSage(): Generator<ForkEffect> {
	yield fork(equallySaga)
	yield fork(userListSaga)
}
