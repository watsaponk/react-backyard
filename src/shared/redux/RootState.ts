import { store } from './Store'

type RootState = ReturnType<typeof store.getState>

export default RootState
