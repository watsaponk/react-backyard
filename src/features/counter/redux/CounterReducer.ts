import { createSlice } from '@reduxjs/toolkit'
import CounterState from './CounterState'

const initialState: CounterState = {
	value: 0,
}

export const slice = createSlice({
	name: 'couter',
	initialState,
	reducers: {
		increase: state => {
			state.value += 1
		},
		decrease: state => {
			state.value -= 1
		},
	},
})

export const { increase, decrease } = slice.actions

export default slice.reducer
