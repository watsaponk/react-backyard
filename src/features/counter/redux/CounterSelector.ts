import { createSelector } from '@reduxjs/toolkit'
import RootState from '../../../shared/redux/RootState'
import CounterState from './CounterState'

const counterSelector = (state: RootState) => state.counter

const valueSelector = createSelector(counterSelector, (state: CounterState) => state.value)

export default valueSelector
