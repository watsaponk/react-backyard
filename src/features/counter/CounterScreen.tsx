import React from 'react'
import { Text, SafeAreaView, View, Button } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { RootState } from '../../shared/Store'
import { decrease, increase } from './redux/CounterReducer'

export default function CounterScreen(): React.ReactElement {
	const count = useSelector((state: RootState) => state.counter.value)
	const dispatch = useDispatch()

	return (
		<SafeAreaView style={{ padding: 16, flex: 1 }}>
			<View style={{ flexDirection: 'row', alignItems: 'center', justifyContent: 'space-between' }}>
				<Button testID='button_minus' title='MINUS' onPress={() => dispatch(decrease())} />
				<Text testID='text_count'>{count}</Text>
				<Button testID='button_plus' title='PLUS' onPress={() => dispatch(increase())} />
			</View>
		</SafeAreaView>
	)
}
