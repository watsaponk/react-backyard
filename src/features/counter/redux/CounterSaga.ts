import { call, ForkEffect, put, takeLatest } from 'redux-saga/effects'
import GetMagicNumber from '../../service/GetMagicNumber'
import { initialize, initSuccess } from './CounterReducer'

function* handleInitialize() {
	const response: number = yield call(GetMagicNumber)
	yield put(initSuccess(response))
}

export default function* counterSaga(): Generator<ForkEffect> {
	yield takeLatest(initialize.type, handleInitialize)
}
