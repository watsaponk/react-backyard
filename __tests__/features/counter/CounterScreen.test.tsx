import React from 'react'
import { render, RenderAPI, fireEvent, waitFor } from '@testing-library/react-native'
import { Provider } from 'react-redux'
import { Store } from '@reduxjs/toolkit'
import CounterScreen from '../../../src/features/counter/CounterScreen'
import { increase, decrease, initialize } from '../../../src/features/counter/redux/CounterReducer'
import GetMagicNumber from '../../../src/features/service/GetMagicNumber'
import buildStore from '../../../src/shared/redux/Store'

jest.mock('../../../src/features/service/GetMagicNumber', () =>
	jest.fn().mockImplementation(() => {
		return Promise.resolve(2)
	})
)

const renderScreen = (store: Store): RenderAPI => {
	return render(
		<Provider store={store}>
			<CounterScreen />
		</Provider>
	)
}

describe('CounterScreen', () => {
	test('When component is mount should get magic number', async () => {
		const store = buildStore()
		const spyDispatch = jest.spyOn(store, 'dispatch')
		const { getByTestId } = renderScreen(store)
		const textMagicNumber = getByTestId('text_magic_number')

		await waitFor(() => {
			expect(GetMagicNumber).toHaveBeenCalledTimes(1)
			expect(spyDispatch).toHaveBeenLastCalledWith(initialize())
			expect(textMagicNumber).toHaveTextContent(/^Magic Number 2$/)
			expect(store.getState().counter.magicNumber).toEqual(2)
			spyDispatch.mockClear()
		})
	})

	test('When press PLUS button should dispatch increase action', async () => {
		const store = buildStore()
		const spyDispatch = jest.spyOn(store, 'dispatch')
		const { getByTestId } = renderScreen(store)
		const buttonPlus = getByTestId('button_plus')
		const textCount = getByTestId('text_count')

		fireEvent.press(buttonPlus)

		await waitFor(() => {
			expect(spyDispatch).toHaveBeenCalledWith(increase())
			expect(textCount).toHaveTextContent(/^1$/)
			expect(store.getState().counter.value).toEqual(1)
			spyDispatch.mockClear()
		})
	})

	test('When press MINUS button should dispatch decrease action', async () => {
		const store = buildStore()
		const spyDispatch = jest.spyOn(store, 'dispatch')
		const { getByTestId } = renderScreen(store)
		const buttonMinus = getByTestId('button_minus')
		const textCount = getByTestId('text_count')

		fireEvent.press(buttonMinus)

		await waitFor(() => {
			expect(spyDispatch).toHaveBeenCalledWith(decrease())
			expect(textCount).toHaveTextContent(/^-1$/)
			expect(store.getState().counter.value).toEqual(-1)
			spyDispatch.mockClear()
		})
	})
})
