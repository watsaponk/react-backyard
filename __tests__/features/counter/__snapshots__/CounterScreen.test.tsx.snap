// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterScreen When render on start 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="button_minus"
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          MINUS
        </Text>
      </View>
    </View>
    <Text
      testID="text_count"
    >
      0
    </Text>
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="button_plus"
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          PLUS
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
